lib/utils.js

```diff
@@ -229,6 +229,146 @@ const format = body => {
+ */
+
+/*
+ * Inititalizes reference
+ *
+ * @param {object} types - content types object
+ *
+ *
+ * * @returns {object} - content type and reference data
+ */
+const references = (types) => {
+  const cts = _.cloneDeep(types);
+  const refs = [];
+  const position = {};
+
+  // Saves position of each content type
+  cts.forEach((type, index) => {
+    if (!position.hasOwnProperty(type.id)) {
+      position[type.id] = index;
+    }
+  });
+
+  // Iterates through each content type and attributes
+  cts.map(ct => {
+    ct.attributes.map((attr, index) => {
+      let inputs;
+      let length;
+
+      // Get the number of instances if repeatable
+      if (Array.isArray(attr.inputs)) {
+        inputs = attr.inputs[0];
+        length = attr.inputs.length;
+      }
+      else {
+        inputs = attr.inputs;
+      }
+
+      // Iterate through each inputs
+      Object.keys(inputs).forEach(input => {
+        // Checks if input is a reference
+        if (inputs[input].hasOwnProperty('reference') && inputs[input].reference === true) {
+          // Throw error if contentType is not defined
+          if (!_.get(inputs[input], 'settings.contentType')) {
+            throw new Error('Reference must have a content type');
+          }
+
+          if (!position.hasOwnProperty(inputs[input].settings.contentType)) {
+            throw new Error(`Content Type ${inputs[input].settings.contentType} is not valid`);
+          }
+
+          if (_.get(inputs[input], 'settings.view') === 'radio') {
+            inputs[input].type = 'radio';
+          }
+
+          // Defines and push reference object
+          refs.push({
+            type: ct.id,
+            attr: index,
+            input,
+            length,
+            ct: inputs[input].settings.contentType,
+          });
+        }
+      });
+    });
+  });
+
+  return {
+    cts,
+    references: refs,
+  };
+};
+
+
+/*
+ * Fills in content type data from database
+ *
+ * @param {object} types - content types object
+ * @param {object} type - content type
+ * @param {array} refs - reference data
+ * @param {object} database - database object
+ *
+ *
+ * @returns {object} - content type object with reference data
+ */
+const fill = (types, type, refs, database) => {
+  const ct = _.cloneDeep(type);
+  if (!ct) {
+    return new Promise(resolve => {
+      resolve(ct);
+    });
+  }
+  const ref = refs.filter(reference => {
+    return ct.id === reference.type;
+  });
+
+  return Promise.map(ref, input => {
+    // subquery for select
+    const where = database
+          .max('revision')
+          .from(`content-type--${input.ct}`)
+          .groupBy('id');
+
+    // Get the records from content-type
+    return database
+      .select('id', 'revision', 'value')
+      .from(`content-type--${input.ct}`)
+      .whereIn('revision', where).then(rws => {
+        const rows = routes.identifier(rws, singleItem('id', input.ct, types));
```

Comment:
`singleItem` should be able to be replaced with [`Array.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) to reduce dependency on the actual `utils` file. This should also probably be done before the function call in order to reduce our call stack as it's not dependent on the result of the function it's in